### PR TITLE
Serialize network architecture to protobuf

### PR DESCRIPTION
**Pull request by graham63**
This adds the capability to output network architectures to a protobuf message that's suitable for ingestion. This PR does NOT include:

1. Data readers or associated capabilities
2. Training algorithms or associated capabilities
3. Weights tensor data

### CODE REVIEW

File: src/layers/activations/identity.cpp [hunk 1]
```diff
@@ -47,6 +48,12 @@ setup_distributions(tensor_overlap_constraints &constraints) {
   constraints.mark_equivalent(dx, dy);
 }
 
+template <typename T, data_layout L, El::Device D>
+void identity_layer<T,L,D>::write_specific_proto(lbann_data::Layer& proto) const {
+  proto.set_datatype(proto::ProtoDataType<T>);
+  proto.mutable_identity();
+}
+
```
Comment (benson31) [inline src/layers/activations/identity.cpp:56]: These are in an `LBANN_HAS_DISTCONV` block.

File: src/layers/operator_layer.cpp [hunk 2]
```diff
@@ -43,4 +44,15 @@ namespace lbann {
...
 #include "lbann/macros/instantiate_device.hpp"
 
+template <typename T, typename O, data_layout L, El::Device D>
+void OperatorLayer<T, O, L, D>::write_specific_proto(lbann_data::Layer& proto) const {
+
+  proto.set_datatype(proto::ProtoDataType<T>);
+  auto* msg = proto.mutable_operator_layer();
+  auto* op = msg->add_ops();
+  op->set_input_datatype(proto::ProtoDataType<T>);
+  op->set_input_datatype(proto::ProtoDataType<O>);
```
Comment (benson31) [inline src/layers/operator_layer.cpp:54]:
```suggestion
  op->set_output_datatype(proto::ProtoDataType<O>);
```

File: src/models/model.cpp [hunk 3]
```diff
@@ -2632,11 +2632,40 @@ bool model::load_from_checkpoint_distributed(persist& p)
...
+  //proto.add_callback();
+
+  //proto.set_summarizer();
+
+  for (auto const& l : m_layers)
+  {
+    auto* l_msg = proto.add_layer();
+    l->write_proto(*l_msg);
+  }
+
```
Comment (benson31) [inline src/models/model.cpp:2668]: It looks like this needs to be fleshed out still.

File: src/layers/learning/embedding.cpp [hunk 4]
```diff
@@ -40,7 +40,7 @@ void embedding_layer<TensorDataType,Layout,Device>::fp_compute() {
   const size_t input_size = this->get_input_size();
   const size_t local_mini_batch_size = local_input.Width();
 
-  // Populate output matrix with values from embedding matrix
+  // Populate out put matrix with values from embedding matrix
```
Comment (benson31) [inline src/layers/learning/embedding.cpp:43]:
```suggestion
  // Populate output matrix with values from embedding matrix
```

File: src/layers/transform/reduction.cpp [hunk 5]
```diff
@@ -46,6 +47,23 @@ std::unique_ptr<Layer> build_reduction_layer_from_pbuf(
...
+  switch (m_mode)
+  {
+    case reduction_mode::SUM:
+      msg->set_mode("sum");
+      break;
+    case reduction_mode::AVERAGE:
+      msg->set_mode("mean");
+      break;
+    default:
+      LBANN_ERROR("Invalid reduction mode requested.");
```
Comment (benson31) [inline src/layers/transform/reduction.cpp:63]:
```suggestion
      msg->set_mode("invalid");
```

File: include/lbann/optimizers/#adam.hpp# [hunk 6]
```diff
@@ -0,0 +1,206 @@
+////////////////////////////////////////////////////////////////////////////////
```
Comment (benson31) [inline include/lbann/optimizers/#adam.hpp#:1]: Remove this file

File: include/lbann/optimizers/data_type_optimizer.hpp [hunk 7]
```diff
@@ -116,7 +116,15 @@ class data_type_optimizer
 
   ///@}
 
+  /** @brief Add optimizer to prototext */
+  void write_proto(lbann_data::Optimizer& proto) const
+  { this->write_specific_proto(proto); }
```
Comment (benson31) [inline include/lbann/optimizers/data_type_optimizer.hpp:121]: Discussed offline.

File: include/lbann/utils/protobuf/impl.hpp [hunk 8]
```diff
@@ -41,6 +41,28 @@
...
+{
+  /** @todo Change to Assign if older versions of protobuf are no
+   *  longer supported.
+   */
+  field.Clear();
+  field.Add(begin(values), end(values));
+}
+
+template <typename ContainerT>
+std::string to_space_sep_string(ContainerT values)
```
Comment (benson31) [inline include/lbann/utils/protobuf/impl.hpp:57]:
```suggestion
std::string to_space_sep_string(ContainerT const& values)
```

File: include/lbann/weights/initializer.hpp [hunk 9]
```diff
@@ -75,6 +79,15 @@ class data_type_weights_initializer
   /** Initialize entries in a weights matrix. */
   virtual void fill(AbsDistMatrixType& matrix) = 0;
 
+  /** @brief Add initializer to prototext */
+  void write_proto(lbann_data::Initializer& proto) const
+  { this->write_specific_proto(proto); }
```
Comment (benson31) [inline include/lbann/weights/initializer.hpp:84]: Discussed offline

File: include/lbann/utils/protobuf/impl.hpp [hunk 10]
```diff
@@ -120,6 +121,27 @@ auto get_repeated_field_ref(google::protobuf::Message const& msg,
...
+
+template <typename ContainerT>
+std::string lbann::protobuf::to_space_sep_string(ContainerT const& values)
+{
+  std::string combined;
+  for (auto const& value : values)
+    combined += (value + " ");
+
+  return combined;
+}
```
Comment (benson31) [inline include/lbann/utils/protobuf/impl.hpp:143]:
```suggestion
/** @brief Stringify the container with spaces between elements.
 *
 *  The `value_type` of the container must support stream output
 *  (e.g., `vector` or `set`, but not `map`).
 */
template <typename ContainerT>
std::string lbann::protobuf::to_space_sep_string(ContainerT const& values)
{
  if (values.empty())
    return "";
  std::ostringstream oss;
  oss << *cbegin(values);
  for (auto i = next(cbegin(values)); i != cend(values); ++i)
    oss << " " << *i;
  return oss.str();
}
```
We should be careful to allow containers of streamable non-strings (your impl will only work for strings). The stream operator is easier to extend than `to_string` for user types.

File: include/lbann/utils/protobuf/impl.hpp [hunk 11]
```diff
@@ -120,6 +121,27 @@ auto get_repeated_field_ref(google::protobuf::Message const& msg,
 } // namespace protobuf
 } // namespace lbann
 
+template <typename T, typename ContainerT>
+void lbann::protobuf::assign_to_repeated(google::protobuf::RepeatedField<T>& field,
+                        ContainerT const& values)
+{
+  /** @todo Change to Assign if older versions of protobuf are no
+   *  longer supported.
+   */
```
Comment (benson31) [inline include/lbann/utils/protobuf/impl.hpp:130]: `RepeatedField<T>::Assign` was added in [Protobuf v3.16.0](https://github.com/protocolbuffers/protobuf/releases/tag/v3.16.0).

File: include/lbann/layers/transform/slice.hpp [hunk 12]
```diff
@@ -138,6 +144,14 @@ class slice_layer : public data_type_layer<TensorDataType> {
...
 // =========================================================
 
+template <typename T, data_layout L, El::Device D>
+void slice_layer<T,L,D>::write_specific_proto(lbann_data::Layer& proto) const {
+  proto.set_datatype(proto::ProtoDataType<T>);
+  auto* msg = proto.mutable_slice();
+  msg->set_axis(m_slice_dim);
+  protobuf::assign_to_repeated(*msg->mutable_slice_points(), m_slice_points);
+}
+
```
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/layers/transform/slice.hpp:154]: @bvanessen What are your thoughts on this?

File: src/models/model.cpp [hunk 13]
```diff
@@ -2632,11 +2632,57 @@ bool model::load_from_checkpoint_distributed(persist& p)
...
+  switch (layout) {
+  case data_layout::DATA_PARALLEL:
+    proto.set_data_layout("data_parallel");
+    break;
+  case data_layout::MODEL_PARALLEL:
+    proto.set_data_layout("model_parallel");
+    break;
+  default:
+    LBANN_ERROR("Invalid data layout");
+  }
```
Comment (benson31) [inline src/models/model.cpp:2659]: Remove from here. This allows a user to set a default data layout for their model, which can be a useful short-hand. However, by the time this function can be called, each layer in the model will have established its data layout, so a default is not needed anymore.

File: include/lbann/utils/protobuf/decl.hpp [hunk 14]
```diff
@@ -205,6 +205,20 @@ std::string serialize(google::protobuf::Message const& msg);
...
+
+/** @brief Assign a range of values to a repeated protobuf field. */
+template <typename T, typename ContainerT>
+void assign_to_repeated(google::protobuf::RepeatedField<T>& field,
+                        ContainerT const& values);
+/** @brief Concatenate a container of strings into a single string.
+ *  Elements are space separated.
+ */
+template <typename ContainerT>
+std::string to_space_sep_string(ContainerT const& values);
```
Comment (benson31) [inline include/lbann/utils/protobuf/decl.hpp:220]: Please add testing for these functions to [the Protobuf utilities tests](https://github.com/LLNL/lbann/blob/develop/src/utils/unit_test/protobuf_utils_test.cpp).

**Review by benson31 — APPROVED** [review]
Looks good.

Prior to merging, can you please add a brief description and mark the PR as "Ready for Review"?

**Comment by benson31** [comment]
Also, please update the release notes. Thanks!

**Review by benson31 — APPROVED** [review]

File: include/lbann/layers/transform/slice.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (bvanessen) [reply include/lbann/layers/transform/slice.hpp:154]: I think that we just write out / hardcode the value that we learned from the input layer.  So it is no longer an exact replica of the input prototext, but is a correct implementation of the model.

**Review by bvanessen — APPROVED** [review]

**Review by benson31 — APPROVED** [review]
LGTM

**Pull request merged**